cpp/src/round_robin/round_robin.cu

```diff
@@ -40,6 +40,122 @@
+  
+template<typename T>
+using VectorT = rmm::device_vector<T>;  
+/**
+ * @brief Handles the "degenerate" case num_partitions >= num_rows.
+ *
+ * Specifically,
+ * If num_partitions == nrows:
+ * Then, offsets = [0..nrows-1]
+ * gather_row_indices = rotate [0..nrows-1] right by start_partition positions;
+ *
+ * If num_partitions > nrows:
+ * Then, let:
+ * dbg = generate a directed bipartite graph with num_partitions nodes and nrows edges, 
+ * so that node j has an edge to node (j+start_partition) % num_partitions, for j = 0,...,nrows-1;
+ *
+ * transpose_dbg = transpose graph of dbg; (i.e., (i -> j) edge in dbg means (j -> i) edge in transpose);
+ *
+ * (offsets, indxs) = (row_offsets, col_indices) of transpose_dbg;
+ * where (row_offsets, col_indices) are the CSR format of the graph;
+ *
+ * @Param[in] input The input table to be round-robin partitioned
+ * @Param[in] num_partitions Number of partitions for the table
+ * @Param[in] start_partition Index of the 1st partition
+ * @Param[in] mr Device memory allocator
+ * @Param[in] stream cuda stream to execute on
+ *
+ * @Returns A std::pair consisting of an unique_ptr to the partitioned table and the partition offsets for each partition within the table
+ */
+std::pair<std::unique_ptr<cudf::experimental::table>,
+          std::vector<cudf::size_type>>
+degenerate_partitions(cudf::table_view const& input,
+                      cudf::size_type num_partitions,
+                      cudf::size_type start_partition,
+                      rmm::mr::device_memory_resource* mr,
+                      cudaStream_t stream)
+{
+  auto nrows = input.num_rows();
+    
+  //iterator for partition index rotated right by start_partition positions:
+  //
+  auto rotated_iter_begin =
+    thrust::make_transform_iterator(thrust::make_counting_iterator<cudf::size_type>(0),
+                                    [num_partitions, start_partition] __device__ (auto index){
+                                      return (index + num_partitions - start_partition) % num_partitions;
+                                    });
+
+  if( num_partitions == nrows ) {
+    VectorT<cudf::size_type> partition_offsets(num_partitions, cudf::size_type{0});
+    auto exec = rmm::exec_policy(stream);
+    thrust::sequence(exec->on(stream),
+                     partition_offsets.begin(), partition_offsets.end());
+
+    auto uniq_tbl = cudf::experimental::detail::gather(input,
+                                                       rotated_iter_begin, rotated_iter_begin + nrows,//map
+                                                       false, false, false,
+                                                       mr,
+                                                       stream);
+
+    auto ret_pair =
+      std::make_pair(std::move(uniq_tbl), std::vector<cudf::size_type>(num_partitions));
+
+    cudaMemcpy(ret_pair.second.data(), partition_offsets.data().get(), sizeof(cudf::size_type)*num_partitions, cudaMemcpyDeviceToHost);
```

Comment:
I think this should be cudaMemcpyAsync, with the stream parameter. You can add cudaStreamSynchronize() afterwards if the API is meant to be synchronous.
Also should be in CUDA_TRY macro.

Reply:
Done.